transcom/mymove
# MB-6481 Sign user out after being idle

## Description

In May 2018, we implemented a feature that would warn the user that
they have been idle, and if they remain idle, they will be signed out.
We used `react-idle-timer` version 3.0 for this functionality, and it
worked. But then in November 2019, we [upgraded](https://github.com/transcom/mymove/commit/58e1e93604637bf7fcfa70b358d7127254182c6b) `react-idle-timer`
from 3.0.0 to 4.2.12, and because we didn't have any tests for this
feature, we didn't notice that it broke.

Version 4.0 introduced new syntax, and so the React component never
detected an idle state. Even though `react-idle-timer` was no longer
working, we should have been able to rely on the session cookie to
expire as a fallback, so that even though the user wasn't returned to
the sign in page, if they refreshed the page, they should have been
signed out. However, the original implementation added an interval
that would ping the `/internal/swagger.yaml` endpoint every 10 minutes.
This unfortunately kept the session alive, well past the required
15-minute expiration for an idle session. Back then, I don't think we
set a lifetime for the session, and so until we moved to server-side
sessions in June 2020, this would keep the session alive forever.

In this PR, I fixed the syntax issues (`activeAction` -> `onAction`,
and `idleAction` -> `onIdle`), and I also made some improvements:

- The internal endpoint is no longer pinged automatically by us. It
only gets pinged if the user becomes active again.
- To reduce the amount of requests to the internal endpoint, I limited
the events that make the user active to pressing a key or touching/
clicking on the page. By default `react-idle-timer` looks for several
more events, such as a mouse move, but making a request every time the
mouse moves is not efficient. Also, most websites that implement idle
timeouts display a modal or some other prompt that the user must click
on to remain active, so I chose events that are similar to that.
- I renamed the constants to better represent what they mean, as
opposed to naming them after specific time amounts. I also split out
the minutes, seconds, and milliseconds to make it easier to change
values for testing without having to do math in your head.
- I added a countdown that shows how much time is left before you get
signed out. Previously, the alert only said that you will be logged
out "shortly".
- I removed the `isProduction` condition to make it easier to test
locally.
- I fixed the relationship between the different time constants so
that the maximum idle time is respected. Before, the user would end
up being idle for an extra minute because the `fifteenMinutesInMilliseconds`
constant is what determined when the inactivity alert first appeared,
and then the user was given an additional minute via `oneMinuteInMilliseconds`
before being signed out.

## Reviewer Notes

Notes for improvement:

- I was hoping to be able to set the `maxIdleTime` to point to the
`SESSION_IDLE_TIMEOUT_IN_MINUTES` env var, so that if we ever change
the default, we wouldn't have to remember to change it here as well.
It looks like we don't have access to the env var as is, but we would
need to define a new `REACT_APP_SESSION_IDLE_TIMEOUT_IN_MINUTES` env
var.

- We define an alert that tells the user that they have been signed
out due to inactivity, but it only shows up right before they get
signed out. It doesn't remain visible on the sign in screen, which is
where the alert would be most useful. I couldn't figure out how to
do that.

How to test locally:

- Change `maxIdleTimeInSeconds` to `10`
- Change `maxWarningTimeBeforeTimeoutInSeconds` to `5`
- Sign in to `milmovelocal` or `officelocal` with login.gov (not devlocal) and remain idle
- [ ] Verify that after 5 seconds, the inactivity alert appears, and
starts counting down from 10 seconds
- [ ] Verify that after the countdown, you are signed out and returned
to the sign in screen
- Sign back in, and after the alert appears and the countdown begins,
click anywhere on the screen before the time is up
- [ ] Verify that the alert disappears, and that a network request is
made to `/internal/swagger.yaml` every time you click on the screen
- Remain idle again
- [ ] Verify that after 5 seconds of idle time, the inactivity alert
appears, and starts counting down from 10 seconds, and that it only
lasts 10 seconds
- Sign back in
- After the alert appears, move the mouse on the screen, but don't
click it and don't press any keyboard keys
- [ ] Verify that mouse moves do not trigger any network requests, and
that you remain idle
- Change `onActive` and `onIdle` in the React component to `activeAction`
and `idleAction` respectively. This should disable the functionality.
- Add `export SESSION_IDLE_TIMEOUT_IN_MINUTES=1` to your `.envrc.local`
- Stop the server and client
- Run `direnv allow` in both terminal tabs
- Run `make server_run` and `make client_run`
- Remain idle for at least 1 minute
- Refresh the web page
- [ ] Verify that you are signed out and back on the sign in screen


## Setup

Add any steps or code to run in this section to help others prepare to run your code:

```sh
make server_run
make client_run # in another tab
```

## Code Review Verification Steps


 [Querying the Database Safely](https://github.com/transcom/mymove/wiki/Backend-Programming-Guide#querying-the-database-safely)
 have been satisfied.
* Any new migrations/schema changes:

### References
[Jira story](https://dp3.atlassian.net/browse/MB-6481)
